File: src/main/java/org/khun/leetcode/ArrayAndString/GasStation.java
package org.khun.leetcode.ArrayAndString;

public class GasStation {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int totalGas = 0;
        int totalCost = 0;
        int tank = 0;
        int start = 0;

        for (int g : gas) {
            totalGas += g;
        }
        for (int c : cost) {
            totalCost += c;
        }
        if (totalGas < totalCost) return -1;

        for (int i = 0; i < gas.length; i++) {
            tank += (gas[i] - cost[i]);
            if (tank < 0) {
                tank = 0;
                start = i + 1;
            }
        }
        return start;
    }

    public static void main(String[] args) {
        GasStation gs = new GasStation();

        int[] gas1 = {1, 2, 3, 4, 5};
        int[] cost1 = {3, 4, 5, 1, 2};
        System.out.println(gs.canCompleteCircuit(gas1, cost1));

        int[] gas2 = {2, 3, 4};
        int[] cost2 = {3, 4, 3};
        System.out.println(gs.canCompleteCircuit(gas2, cost2));

    }
}

File: src/main/java/org/khun/leetcode/ArrayAndString/GasStation.kt
package org.khun.leetcode.ArrayAndString

fun canCompleteCircuit(gas: IntArray, cost: IntArray): Int {
    var totalGas = gas.sum()
    var totalCost = cost.sum()
    var tank = 0
    var start = 0

    if (totalGas < totalCost) return -1
    for (i in 0 until gas.size) {
        tank += (gas[i] - cost[i])
        if (tank < 0) {
            tank = 0
            start = i + 1
        }
    }
    return start
}

fun main() {
    val gas1 = intArrayOf(1, 2, 3, 4, 5)
    val cost1 = intArrayOf(3, 4, 5, 1, 2)
    println(canCompleteCircuit(gas1, cost1))

    val gas2 = intArrayOf(2, 3, 4)
    val cost2 = intArrayOf(3, 4, 3)
    println(canCompleteCircuit(gas2, cost2))
}